### PR TITLE
Add another case we need to split GEMM

### PR DESCRIPTION
**Pull request by qti-yuduo**
### Description
Split when bias is an intermediate (NATIVE) tensor produced by another op.

### Motivation and Context

1. ORT's `MatMulAddFusion`** (runs before QNN EP sees the graph) fuses `MatMul + Add → Gemm`. This is a general ORT graph optimization that fires whenever it sees the pattern — it doesn't know or care what the Add's second input is.

2. **The problematic pattern** — CLIP-style text projection averaging:
   ```
   mm1 = MatMul(A1, W)
   mm2 = MatMul(A2, W)
   add1 = mm1 + mm2          ← MatMulAddFusion fires: becomes Gemm(A1, W, C=mm2)
   add2 = add1 + mm3         ← MatMulAddFusion fires: becomes Gemm(A1(?), W, C=add1)
   ```
   After fusion, the `C` (bias) input to Gemm is `mm2` or `add1` — an **intermediate NATIVE tensor**, not a constant or graph input.

3. **QNN `FullyConnected` constraint**: QNN's FC op requires the bias to be `STATIC` (initializer/constant) or `APP_WRITE` (graph input). It **cannot accept a `NATIVE` tensor** (intermediate output of another op) as bias. This would cause a validation/compilation failure.

4. **The fix**: When the bias is `NATIVE`, decompose `Gemm → FC(A, W) + ElementWiseAdd(fc_output, C)`. The Add op has no such restriction on input types.

Fixes following issue: 
```
Node(s) placed on [CPUExecutionProvider]. Number of nodes: 17
DequantizeLinear (QcQuantizeOp_model.box_head.conv1_ctr.0.weight_dq)
DequantizeLinear (QcQuantizeOp_model.box_head.conv1_offset.0.weight_dq)
DequantizeLinear (QcQuantizeOp_model.box_head.conv1_size.0.weight_dq)
DequantizeLinear (QcQuantizeOp_model.box_head.conv5_offset.weight_dq)
DequantizeLinear (QcQuantizeOp_val_176_dq)
DequantizeLinear (QcQuantizeOp_val_22_dq)
DequantizeLinear (QcQuantizeOp_val_26_dq)
Gemm (node_MatMul_21/MatMulAddFusion)
Gemm (node_MatMul_23/MatMulAddFusion)
Gemm (node_MatMul_131/MatMulAddFusion)
DequantizeLinear (node_MatMul_21/MatMulAddFusion_bias_dq)
DequantizeLinear (node_MatMul_23/MatMulAddFusion_bias_dq)
DequantizeLinear (node_MatMul_131/MatMulAddFusion_bias_dq)
DequantizeLinear (node_Conv_145_bias_dq)
DequantizeLinear (node_Conv_137_bias_dq)
DequantizeLinear (node_Conv_141_bias_dq)
DequantizeLinear (node_conv2d_11_bias_dq)
```

### CODE REVIEW

File: onnxruntime/test/providers/qnn/gemm_test.cc [hunk 1]
```diff
@@ -496,6 +496,68 @@ TEST_F(QnnHTPBackendTests, Gemm_TransAB_Dynamic_B_And_Bias) {
...
+TEST_F(QnnHTPBackendTests, GemmFromMatMulAddNonStaticBias) {
+  ProviderOptions provider_options;
+  provider_options["backend_type"] = "htp";
+  provider_options["offload_graph_io_quantization"] = "0";
+
+  RunQnnModelTest(BuildGemmFromMatMulAddTestCase(/*K=*/4, /*N=*/3),
+                  provider_options,
+                  /*opset=*/18,
+                  ExpectedEPNodeAssignment::All,
+                  /*fp32_abs_err=*/1e-2f);
```
Comment (quic-ykota) [inline onnxruntime/test/providers/qnn/gemm_test.cc]: Why is the accuracy threshold kept so high? I see couple of other tests using 1e-5f. Could we use that here?

**Review by quic-ykota — COMMENTED** [review]
I wish there was a way to subscribe to ORT when to fuse and when not to fuse, which would avoid a lot of unnecessary logic in EP to undo what ORT has done. But, this change looks good to me.

File: onnxruntime/test/providers/qnn/gemm_test.cc (hunk 1 above)
Comment (qti-yuduo) [reply onnxruntime/test/providers/qnn/gemm_test.cc]: Linux HTP simulator gives slightly higher error in this case. I do see Windows/Android not complaining even with default tolerance which is a good signal.

**Comment by qti-yuduo** [comment]
> I wish there was a way to subscribe to ORT when to fuse and when not to fuse, which would avoid a lot of unnecessary logic in EP to undo what ORT has done. But, this change looks good to me.

Yeah the flexibility would be better. However, note in this case, `MatMulAddFusion` generally helps, thus, we do not actually want to disable entire pass. Undo partial corner cases probably the best thing we could do here.

File: onnxruntime/test/providers/qnn/gemm_test.cc (hunk 1 above)
Comment (quic-ykota) [reply onnxruntime/test/providers/qnn/gemm_test.cc]: Is it possible to change the tolerance for Windows/Android to be 1e-5f and keep it as 1e-2 for linux? This way we can catch any regressions in the future. Keeping it at 1e-2 might mask subtle errors

Comment (qti-yuduo) [reply onnxruntime/test/providers/qnn/gemm_test.cc]: We can do macro I guess, not elegant but doable.

Comment (qti-yuduo) [reply onnxruntime/test/providers/qnn/gemm_test.cc]: Added. Let's see if it passes.

**Review by quic-ykota — APPROVED** [review]

**Review by qti-chuteng — APPROVED** [review]

File: onnxruntime/test/providers/qnn/gemm_test.cc (hunk 1 above)
Comment (qti-yuduo) [reply onnxruntime/test/providers/qnn/gemm_test.cc]: Default still not good enough.. one more try.

**Review by quic-muchhsu — APPROVED** [review]
